/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max = 0;
    public int averageOfSubtree(TreeNode root) {
        dfs(root);
        return max;
    }

    private int[] dfs(TreeNode root) {
        if (root == null) return new int[]{0, 0};

        int[] left = dfs(root.left);
        int[] right = dfs(root.right);

        int c = left[0] + right[0];
        int sum = left[1] + right[1];
        int avg = (sum + root.val) / (c + 1);
        if (avg == root.val) max += 1;
        return new int[]{c + 1, sum + root.val}; 
    }
}